Added module for SoftLayer

mcltn opened the pull request:
This module allows the functionality of building and canceling SoftLayer virtual instances. This module requires the SoftLayer library to work. The parameters allow you to build a single instance as well as specify a quantity of instances to build in addition to the various options for the virtual instances themselves.

robynbergeron commented:
Hi @mcltn --

Thanks for submitting this new module to Ansible Extras! This module is now in community review, a process that is open to all Ansible users. In order for this module to be approved, it must gain the following votes:

“works_for_me”: If you have tested the module thoroughly, including testing of all of the module’s options, and if the module works for you, please add “works_for_me” in the comments.

“passes_guidelines”: If you have gone through the module guidelines and the module meets all of the requirements, please add “passes_guidelines” in the comments. Guidelines are available here: http://docs.ansible.com/developing_modules.html#module-checklist

“needs_revision”: If the module fails to work for you, or if it doesn’t meet guidelines, please add “needs_revision” in the comments with details about what needs to be fixed.

When a module has both “works_for_me” and “passes_guidelines” tags, we will promote the module for inclusion in Ansible Extras.  At this point, you will be expected to maintain the module by fixing bugs and evaluating pull requests in a timely manner. 

Thanks again for submitting your Ansible module!

sp4ceb4r commented:
"passes_guidelines"

Side Note: While PEP8 is not required the use of spaces instead of tabs would make any potential Pull Requests easier as spaces seem to be the more default option.

bcoca commented:
while we don't force PEP8, tabs as indentation do create issues, so please convert those to spaces.

mcltn commented:
All of the tabs contained in the code have been converted to spaces.

coreylane commented:
works_for_me

bcoca commented on cloud/softlayer/softlayer.py:
```diff
@@ -0,0 +1,428 @@
+description:
+  - Creates or cancels SoftLayer instances. When created, optionally waits for it to be 'running'.
+options:
+  hostname:
```
the docs need to be valid YAML, the indentation is off in many places

mcltn commented:
Issues with spacing within documentation yaml has been corrected.

bcoca commented on cloud/softlayer/softlayer.py:
```diff
@@ -0,0 +1,428 @@
+'''
+
+import time
+from ansible.module_utils.facts import *
```
not sure why you have facts imported here

bcoca commented on cloud/softlayer/softlayer.py:
```diff
@@ -0,0 +1,428 @@
+
+def create_virtual_instance(module):
+
+  hostname = module.params.get('hostname')
```
most of these are just then reassigned down below, wouldn't it just be simpler to directly use the module.params?

bcoca commented:
also we would appreciate it if you could flatten this into a single commit

mcltn commented:
Flattened to a single commit.

kbroughton commented:
Looks good.  My one nit would be that softlayer.py is the name of the dynamic inventory script as well.
Perhaps following the lead of the ec2 modules we could call the sl creation
sl.py, thus allowing sl_tag, sl_volume etc to come in the future.  Names will become long if all modules have to be pre-pended by softlayer in full.  ec2 modules for reference: http://docs.ansible.com/ansible/list_of_cloud_modules.html

mcltn commented:
yep, been thinking about that as well. Going to rename to your suggestion.

kbroughton commented:
Testing now.  I'm seeing something odd.  Passing in an int for cpus works fine, but passing a variable that resolves to an int fails.  Could you try passing in a var for cpus and see if you can reproduce?

failed: [localhost -> 127.0.0.1] => (item={'count': 3, 'datacenter': 'dal06', 'hostname': u'worker', 'disks': [25, 100], 'tags': ['mesos_slave', 'hdfs_datanode', 'consul_client'], 'disk_info': [{'device': 'xvda', 'size': 25}, {'device': 'xvdc', 'size': 100}, {'device': 'xvdd', 'size': 200}], 'cpus': 1, 'local_disk': False}) => {"failed": true, }}
msg: value of cpus must be one of: 1,2,4,8,16, got: 1

as you can see, item.cpus is resolving to an int (no quotes around the 1 in 'cpus': 1)
  cpus: "{{item.cpus }}"        - fails when item.cpus = 1
  cpus: 1                              - works fine

mcltn commented:
This throws the error due to the "choices" option in the parameter. Only certain cpu sizes are allowed and the array is of ints.

kbroughton commented:
Docs don't specify if ssh_keys: var takes a keyname or id.  Name doesn't work, I have to use the key id found by
slcli sshkey list
:........:.............................:.................................................:.......:
:   id   :            label            :                   fingerprint                   : notes :
:........:.............................:.................................................:.......:
: 10623  :         dummy        : a3*****:b3 :   -   :

It would be nice to do id lookup of ssh_key names but i wouldn't hold up this pr over it.  Just clarify in the docs.

mcltn commented:
Specified in the docs that VLAN and SSH Keys are by Id.

kbroughton commented:
@mcltn 
Still trying to figure out why i can't pass variables to cpus: and memory:

Here is my debug using a jinja2 test to make sure it is a number
- debug: msg="{{item.cpus }} {{item.cpus is number}}"
  with_items: sl_requested_hosts

where 
sl_requested_hosts:
  cpus: 1

returns  
}
    "msg": "1 True"
}

So it seems I am passing in an int and it is in the list of allowed values.
Other modules take variables no problem.  Am i missing something?

mcltn commented:
looks like adding type='int' to the params fixes the issue. Adding to both cpus and memory.

kbroughton commented:
for consistency with ec2 module, could you move the wait logic to its own top-level arg rather than putting it into state?  I wasn't clear if state = wait would create, then wait, or just wait assuming it was already created.

ec2 docs:
parameter           required            default             choices              comments
wait                    no              no                    yes, no              wait for the instance to be 'running'  
                                                                                                     before returning. Does not wait for 
                                                                                                     SSH, see 'wait_for' example for details.
wait_timeout
                             no                 300                                         how long before wait gives up, in 
                                                                                                    seconds

mcltn commented:
I'm hesitant to do it this way since some users, as well as myself, would want to potentially run provisioning in parallel, but I guess it could place an order in the first play, say for web heads, then when running a second play, say for a db, this step could have a wait flag. It would still require a tag or something to use to know what instances to wait for. Other thoughts?

kbroughton commented:
@mcltn I'll have to check but I think the way "wait" works with ec2 is that all jobs are launched in parallel, but all jobs must complete before moving on to the next task.  I see no use case to waiting for each to complete serially.  You would use the "serial" keyword for that i guess.

kbroughton commented:
when I add a list of tags rather than a single tag I get "array" as the name of the tag rather than a list of the tags.

tag: [ tag1, tag2, tag3 ]

Output: 

```
"status": {
      "keyName": "ACTIVE", 
      "name": "Active"
    }, 
    "tagReferences": [
      {
        "empRecordId": "", 
        "id": 9756287, 
        "resourceTableId": 13964399, 
        "tag": {
          "name": "array"
        }, 
        "tagId": 333695, 
```

I even tried changing to
      tag=dict(type='list'),
from tag=dict()
in
def main():
  module = AnsibleModule(

mcltn commented:
The SoftLayer python library takes tags in a string with tags being separated by commas within the string. This was setup to work the same. Probably just need to update the docs to reflect the proper use. Also looking into doing some refactoring for the wait suggestion.

kbroughton commented:
cool.  I discovered this last night.  I can submit a PR that would allow
the module to ingest either comma delimited string or an ansible list.
 ec2.py accepts an ansible list, so again this is for consistency.  And
again for consistency with ec2, maybe you could make tags an alias for tag.

tag=dict(aliases=['tags'])

On Fri, Nov 20, 2015 at 6:30 AM, matt colton notifications@github.com
wrote:

> The SoftLayer python library takes tags in a string with tags being
> separated by commas within the string. This was setup to work the same.
> Probably just need to update the docs to reflect the proper use. Also
> looking into doing some refactoring for the wait suggestion.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible-modules-extras/pull/1020#issuecomment-158385188
> .

## 

kesten broughton
devops @ cognitivescale

mcltn commented:
Made modifications to use wait logic within the creation of virtual machines. Optional parameters wait and wait_timeout, timeout defaults to 600 and wait is True by default. I also modified the tag param to tags since this was what the Softlayer python library is using also.

mcltn commented:
Just noticed an issue with wait and multiple tags. Should it just default to the first tag in a list for using for the wait? Thoughts?

kbroughton commented:
Regarding tags, I think exact match is most logical.  You could have an "any_tag" option which would switch between the two.  
But i would be very careful about cancelling with tags instead of instance ids.  Right now when I call list_instances(tags="tag1,tag2") i get returned instances that have neither tag1 nor tag2.  I'm running more tests, but it seems like searching multiply tagged instances with multiple search tags is broken.
Until we do thorough testing, i think it is unsafe to allow cancel by tag.

mcltn commented:
So it appears that the vsManager.list_instances() when using tags takes a list. I'm converting it to use it as a list, which is actually how it previously was when wait was a separate state, but was encapsulating the string into [] to make an array. Cancel was done the same way.

mcltn commented:
Also, I noticed the needs revision label had been removed earlier on some fixes/changes, but can you go ahead and put one back until we agree it is ready?

mcltn commented:
Latest has updates for proper use of tags per SoftLayer python library and VS Manager. Updated example section to reflect use of tags as well, between string and list use.

code-noggin commented:
You have an option to specify the private_vlan but do not have one for public_vlan. Would you mind including a public_vlan option same as private_vlan?

robynbergeron commented:
@mcltn -- sorry, just noticed the request to add the needs revision label. Looks like you've been doing some work since that point though -- is it now ready for review?

mcltn commented:
No worries Robyn, I do want to make one more edit to the code around the use of the wait parameters. It needs to only wait for the instances in which it was creating vs using the tag specified. This should be updated and tested soon.

mcltn commented:
Updates completed to "wait" logic to only look at instances that have been created in current process. Using "wait" will provide IP details. Should possibly also add ability to only wait for IP, since IP is not given on return of instances initially and when not waiting.

mcltn commented:
All changes have been completed and tested.

gregdek commented:
@mcltn thanks for all your work here.

@kbroughton thanks for your extensive feedback. Since we've had +1s from other folks, I'd like to wait for your feedback. If you think it's ready to go, give us a "shipit".

gregdek commented:
Actually, it's time to make a judgement call here. This has a lot of +1s, and the maintainer is clearly responsive, even if there may still be minor issues. Therefore, we're going to mark this as "shipit". Thanks @mcltn for all the hard work, and thanks in advance for maintaining this awesome module. :)

farioas commented:
Status?

mcltn commented:
No changes are being made currently or anticipated just yet. Just waiting to get pulled into the main repo. Thanks @gregdek for helping get this pushed along. Looking forward to the add, and next features and enhancements to be added to this module.

bcoca commented on cloud/softlayer/sl.py:
```diff
@@ -0,0 +1,422 @@
+      - Instance Id of the virtual instance to perform action option
+    required: false
+    default: null
+  hostname:
```
hostname is duplicated

bcoca commented on cloud/softlayer/sl.py:
```diff
@@ -0,0 +1,422 @@
+    required: false
+    default: null
+  state:
+    description:
```
this should document the choices 'present' and 'absent', also they should be enforced in the module spec in the code.

mcltn commented:
Updated documentation and added the version_added requirement. Also added choices for the present and absent to the state param. CI build started failing on this commit though due to no return being specified, which was new, so added an empty return as well.

mcltn commented:
Thanks for pointing that out @farioas . Now updated and tested with the newer way to setup the SoftLayer client.

gregdek commented:
Putting back into shipit. Thanks @mcltn!

bcoca commented on cloud/softlayer/sl.py:
```diff
@@ -0,0 +1,430 @@
+      - Quantity of virtual instances to provision
+    required: false
+    default: 0
+  increment:
```
this is a bad pattern that crept into other modules, ansible already allows the use of sequences and loops to take care of this

mcltn replied:
Making edits to remove this functionality within the module. Makes better sense to utilize it outside.

bcoca commented on cloud/softlayer/sl.py:
```diff
@@ -0,0 +1,430 @@
+LOCALDISK_SIZES = [25,100,150,200,300]
+SANDISK_SIZES = [10,20,25,30,40,50,75,100,125,150,175,200,250,300,350,400,500,750,1000,1500,2000]
+NIC_SPEEDS = [10,100,1000]
+
```
is there no way to query the softlayer api for this info? it seems that it will need to be updated anytime the options change

mcltn replied:
These could possibly be pulled via Api calls, but are also pretty static. Only one that may change in near future would be data centers.

bcoca commented on cloud/softlayer/sl.py:
```diff
@@ -0,0 +1,430 @@
+    default: 600
+
+requirements:
+    - "python >= 2.6"
```
wouldn't the softlayer python lib also be required?

mcltn replied:
Will add this on next commit.

gregdek commented:
@mcltn A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

mcltn commented:
Planning on making some adjustments, may need some time. Should have updates coming soon. Thanks!

gregdek commented:
@mcltn A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

kbroughton commented:
if @mcltn doesn't respond by end of week, i can try responding to the
action items on the ticket rather than closing it.

On Thu, Mar 3, 2016 at 9:14 AM, Greg DeKoenigsberg <notifications@github.com

> wrote:
> 
> @mcltn https://github.com/mcltn A friendly reminder: this pull request
> has been marked as needing your action. If you still believe that this PR
> applies, and you intend to address the issues with this PR, just let us
> know in the PR itself and we will keep it open pending your changes.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible-modules-extras/pull/1020#issuecomment-191808185
> .

## 

kesten broughton
devops @ cognitivescale

gregdek commented:
@mcltn A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

mcltn commented:
Removed the ability to specify quantity. The example documentation shows how to perform multiple provisions using the loop feature with_items. I'd still like to see a better way to handle the wait flag, since now this would wait for each instance in serial as they build for ones that are flagged. This could cause quite a delay when building a large volume of servers.

gregdek commented:
Thanks @mcltn. Presuming that you still intend to handle diffs from @bcoca above? Leaving in needs_revision for now, but if you feel that it's ready, reply with ready_for_review and we'll put it back into community review.

mcltn commented:
ready_for_review

JulienChampseix commented:
any update ? @bcoca @gregdek

gregdek commented:
Thanks @mcltn for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

darKoram commented:
@mcltn A little late, but shipit!  Sorry, missed your initial request.

rickji commented:
Question: where do I put in the SoftLayer username and APIkey for authentication?

Blake- commented:
Try putting it in ~/.softlayer
Takes the following form, but with real username & key. Also I think the endpoint may differ a bit depending on what you do?

```
[softlayer]
username = SL99999
api_key = 80924t2h4ti2h4t2jh4tk2j4thj2kht4j2kht
endpoint_url = https://api.softlayer.com/xmlrpc/v3.1/
```
